server/core/db/boltdb.go

```diff
@@ -106,23 +114,80 @@ func (b *BoltDB) TryLock(newLock models.ProjectLock) (bool, models.ProjectLock,
 		if err := json.Unmarshal(currLockSerialized, &currLock); err != nil {
 			return errors.Wrap(err, "failed to deserialize current lock")
 		}
+
+		// checking if current lock is with the same PR
+		if currLock.Pull.Num == newLock.Pull.Num {
+			lockAcquired = true
+			return nil
+		}
+
 		lockAcquired = false
+
+		// Enqueuing
+		queueBucket := tx.Bucket(b.queueBucketName)
+		currQueueSerialized := queueBucket.Get([]byte(key))
+		// Queue doesn't exist, create one
+		if currQueueSerialized == nil {
+			newQueue := models.ProjectLockQueue{newLock}
+			newQueueSerialized, err := json.Marshal(newQueue)
+			if err != nil {
+				return errors.Wrap(err, "serializing")
+			}
+			if err := queueBucket.Put([]byte(key), newQueueSerialized); err != nil {
+				return err
+			}
+			enqueueStatus = models.EnqueueStatus{
+				Status:     models.Enqueued,
+				QueueDepth: 1,
+			}
+			return nil
+		}
+		// Queue exists
+		var currQueue models.ProjectLockQueue
+		if err := json.Unmarshal(currQueueSerialized, &currQueue); err != nil {
+			return errors.Wrap(err, "failed to deserialize queue for current lock")
+		}
+
+		// Lock is already in the queue
+		if indexInQueue := currQueue.FindPullRequest(newLock.Pull.Num); indexInQueue > -1 {
+			enqueueStatus = models.EnqueueStatus{
+				Status:     models.AlreadyInTheQueue,
+				QueueDepth: indexInQueue + 1,
+			}
+			return nil
+		}
+		// Not in the queue, add it
+		newQueue := append(currQueue, newLock)
+		newQueueSerialized, err := json.Marshal(newQueue)
+		if err != nil {
+			return err
+		}
+		if err := queueBucket.Put([]byte(key), newQueueSerialized); err != nil {
+			return err
+		}
```

Comment:
Since these blocks will return if an error is not nil, will the function calling this function know to log the error? or should we log it here?

Reply:
Yes, the error is logged by `events/instrumented_project_command_runner.go:77`:
```
{"level":"error","ts":"2023-03-17T15:43:55.479+0100","caller":"events/instrumented_project_command_runner.go:77","msg":"Error running plan operation: acquiring lock: DB transaction failed: custom exception for testing", [..etc]
```

And shown in the PR comment:
<img width="934" alt="image" src="https://user-images.githubusercontent.com/20379416/225938015-2706dc52-36a8-4c1f-83a0-93b814bc4b9f.png">